### PR TITLE
aws - output - try to determine bucket region without a client

### PR DESCRIPTION
**Pull request by ajkerrigan**
Following some recent changes to the way we determine output bucket regions (#7524) we've seen some reports of access issues in certain cross-account/cross-region output scenarios (#7593). @MichaelDavisTSN also raised the issue [in Gitter](https://gitter.im/cloud-custodian/cloud-custodian?at=62fbcb89aa091774297141a2).

In trying to reproduce and work around those issues, I ended up sneaking up on an approach @sean-zou mentioned - grabbing the region from the `x-amz-bucket-region` header of an HTTP HEAD response.

~This wouldn't cover _all_ cases - non-standard partitions like GovCloud come to mind.~(Actually I _think_ it can if we're careful) I'm wondering if we can try the HTTP HEAD request first, then fall back to `get_bucket_location`. And if neither works, give up.

I need to add some tests and let some other folks try this, so leaving it in draft for the moment. But if anyone can poke holes in the  general idea in the meantime, fire away :).

### CODE REVIEW

File: c7n/resources/aws.py
Comment (kapilt) [inline c7n/resources/aws.py:140]: can you ref the link where this suggestion came from?

Comment (ajkerrigan) [reply c7n/resources/aws.py:146]: Sure, good call :+1:

**Comment by ajkerrigan** [comment]
> for partitions I think we should check if the region is in a partition, and if so force the assumption the bucket is in the region, or use the client to determine location.
> 
> I'm still missing where this issue arises, is it solely cross account bucket outputs? or does it happen for single account cross region?

Ah, that might be a smarter way to handle non-standard partitions. Good call. I haven't been able to trigger it for single account cross-region or cross-account in the same region. Only cross-account _and_ cross-region.

File: c7n/resources/aws.py
Comment (thisisshi) [inline c7n/resources/aws.py:588]: region here could be None for us-east-1 fyi, it's probably better to do a try/except on the get_bucket_location call instead if you want to raise

**Comment by PratMis** [comment]
> for partitions I think we should check if the region is in a partition, and if so force the assumption the bucket is in the region, or use the client to determine location.
> 
> I'm still missing where this issue arises, is it solely cross account bucket outputs? or does it happen for single account cross region?

@kapilt this issue seems to be popping up when output buckets are cross region. Doesn't seem to have any dependency with cross account

File: c7n/resources/aws.py
Comment (ajkerrigan) [reply c7n/resources/aws.py:588]: Hmm I was actually thinking about going the other way and ripping the get_bucket_location call out entirely. Between the unintuitive return values (None = 'us-east-1', EU = 'us-west-1) and access issues, HTTP requests seem more reliable. Left this in as a backstop though, because I assumed there was a case I was missing :).

Valid point though, if we're going to make the call we should be thorough and catch the "weird" regions...

**Comment by ajkerrigan** [comment]
Ok after a bit of testing whack-a-mole this seems good to go, unless there are any lingering concerns about the flow of HTTP HEAD with a GetBucketLocation fallback.

File: c7n/resources/aws.py
Comment (kapilt) [inline c7n/resources/aws.py:603]: sys exit here feels odd, ie library causing programing exit, versus raised exception

File: c7n/resources/aws.py
Comment (kapilt) [inline c7n/resources/aws.py:147]: this doesn't seem to work for buckets with a '.' in their name.

File: c7n/resources/aws.py
Comment (kapilt) [inline c7n/resources/aws.py:597]: re public attribute to get host client.meta.endpoint_url

File: c7n/resources/aws.py
Comment (ajkerrigan) [reply c7n/resources/aws.py:147]: True... possible workarounds I can see:

- Using HTTP rather than HTTPS
- Ignoring certificate validation
- Switching to a path-style URL (keeping https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/ in mind)

I'm leaning toward that last one. Thanks, good catch!

File: c7n/resources/aws.py
Comment (ajkerrigan) [reply c7n/resources/aws.py:603]: Fair, I switched that out because the exception looked unnecessarily noisy and I saw the exit pattern in a couple other spots. Exception does make more sense though, will switch back.

**Comment by PratMis** [comment]
Hi @ajkerrigan / @kapilt , is there anything else that needs to happen in this PR? If not, can we please merge it? Thanks

**Comment by ajkerrigan** [comment]
> Hi @ajkerrigan / @kapilt , is there anything else that needs to happen in this PR? If not, can we please merge it? Thanks

I've made some updates based on feedback from @thisisshi and @kapilt , a couple high-level things that people might :raised_eyebrow: about though:

- There were a few options to handle the "bucket with dots in the name case". Using path-style URLs seemed like the best of a few suboptimal options, but I could see arguments for others.
- Is the intent and flow clear enough to someone who _wasn't_ just troubleshooting `GetBucketLocation` issues? I'm too biased to know. Of _course_ it makes sense... **to me** :sweat_smile: .

**Comment by kapilt** [comment]
Fwiw, would consider this pr a release blocker.

**Review by thisisshi — APPROVED** [review]
thanks, this lgtm though i would defer to @PratMis if he can validate that it works in his environment as well

**Comment by PratMis** [comment]
Thanks @thisisshi, I'm testing it right now. Let me report back here

**Comment by ajkerrigan** [comment]
Thanks @PratMis for testing and following up here :+1: .

I'm wondering if going forward we might want to use this approach while assembling a bucket, to put a top-level region annotation on a bucket resource (relevant for issues like #5925). It might also be useful to create regional clients for filters/actions (I'm thinking specifically of the `metrics` filter, which can be unreliable if you're running s3 policies from a single region but cloudwatch metrics are split across regions). Those are all separate discussions though.

**Comment by PratMis** [comment]
Yes, afaics policy authors also struggle with S3 bucket regions. Having a top-level region annotation and using that for regional endpoints make a lot of sense to me. I'm kinda wondering if this is the only good way to find out the region of a bucket.

**Pull request merged**
